### PR TITLE
Modularized gaussian noises' dtype.

### PR DESCRIPTION
**Pull request by pandrey-fr**
As a reply to issue tensorflow#30834, I propose this very small fix that enforces the adjustment of gaussian noises' dtype with that of the inputs within tf.keras.layers.GaussianNoise and tf.keras.layers.GaussianDropout.

### CODE REVIEW

**Comment by pandrey-fr** [comment]
I signed it!

**Review by tanzhenyu — COMMENTED** [review]
Can you add a test?

**Comment by pandrey-fr** [comment]
**Test code**

I derived this test script from the example provided by @bluehope, author of issue #30834 which reported the bug my PR fixes.

```python
import tensorflow as tf


def make_model(dtype, gtype='noise'):
    assert dtype in (tf.float32, tf.float64)
    assert gtype in ('noise', 'dropout')
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(10, input_shape=(10,), dtype=dtype))
    if gtype == 'noise':
        gaussian = tf.keras.layers.GaussianNoise(0.0003)
    else:
        gaussian = tf.keras.layers.GaussianDropout(0.1)
    model.add(gaussian)
    return model


if __name__ == '__main__':
    for dtype in (tf.float32, tf.float64):
        for gtype in ('noise', 'dropout'):
            try:
                make_model(dtype, gtype)
            except:
                status = 'failure'
            else:
                status = 'success'
            print(f'Test with dtype {dtype} and gaussian {gtype}: {status}.')
```

---

Now, I ran this test script with tensorflow 2.0-beta1 (with GPU support), but it should be the same with 1.14 (and without GPU).

**Output before the fix**
My first run used TF installed from binary using pip (on a Linux Mint 19.1 system).
```
Test with dtype <dtype: 'float32'> and gaussian noise: success.
Test with dtype <dtype: 'float32'> and gaussian dropout: success.
Test with dtype <dtype: 'float64'> and gaussian noise: failure.
Test with dtype <dtype: 'float64'> and gaussian dropout: failure.
```

**Output after the fix**

My second run used the same TF installation, after replacing the file `tensorflow/python/keras/layers/noise.py` in the local library (as this is a fix to python code, I did not recompile from source).

```
Test with dtype <dtype: 'float32'> and gaussian noise: success.
Test with dtype <dtype: 'float32'> and gaussian dropout: success.
Test with dtype <dtype: 'float64'> and gaussian noise: success.
Test with dtype <dtype: 'float64'> and gaussian dropout: success.
```

**Comment by pandrey-fr** [comment]
@tanzhenyu If this example does not suit you and you would like me to write a test following a specific template, please let me know.

**Comment by tanzhenyu** [comment]
> @tanzhenyu If this example does not suit you and you would like me to write a test following a specific template, please let me know.

Oh I mean can you add a test inside the code base?

**Comment by pandrey-fr** [comment]
@tanzhenyu I guess I can, but what exactly do you want that test to verify? The mere possibility to instantiate layers (in which case I will use what I wrote earlier), or also the behavior of the gaussian noise being added to input tensors (using a seed for deterministic results)?

**Comment by pandrey-fr** [comment]
Furthermore, would you like to be
* a) in a separate file
* b) in a `if __name__ == '__main__'` block at the end of the file
* c) in the docstring (i.e. a doctest)
?

**Comment by tanzhenyu** [comment]
Preferrably under noise_test.py, there is a layer_test util you can use

**Comment by pandrey-fr** [comment]
@tanzhenyu As it turns out, there already is a `noise_test.py` file [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/noise_test.py), although it was oddly not included in either of my local installations.

The test runs before and after my fix, but do not actually check for it (nor for compatibility with various input dtypes in general). I would be happy to change that, but there is literally no included documentation on the test utils (seriously, I love tensorflow, but something ought to be done regarding the python source code's general writing and documenting conventions). Anyway, if you have any documentation material that would help - otherwise I will just tweak around hoping to stumble upon the desired feature ;)

**Comment by pandrey-fr** [comment]
I followed the example of `recurrent_test.py` ([here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/recurrent_test.py)) and added a few tests based on instantiating a small model, compiling it and running a fitting process (on small, zero data) - in the vein of the block of code previously shared here.

I do not know if that suits the current testing practices (if not, let me know how to change this), but it does highlight the issue and the validity of the fix as to it (the two gloat64 tests fail without the fix, whereas they all pass after deploying it).

**Comment by tanzhenyu** [comment]
> @tanzhenyu As it turns out, there already is a `noise_test.py` file [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/noise_test.py), although it was oddly not included in either of my local installations.
> 
> The test runs before and after my fix, but do not actually check for it (nor for compatibility with various input dtypes in general). I would be happy to change that, but there is literally no included documentation on the test utils (seriously, I love tensorflow, but something ought to be done regarding the python source code's general writing and documenting conventions). Anyway, if you have any documentation material that would help - otherwise I will just tweak around hoping to stumble upon the desired feature ;)

I thought that's exactly what I commented above: "Preferrably under noise_test.py, there is a layer_test util you can use" :-)

From a TDD principle, you could 1) add a test which should fail, 2) add your code change to make the test succeeds

**Comment by pandrey-fr** [comment]
> I thought that's exactly what I commented above: "Preferrably under noise_test.py, there is a layer_test util you can use" :-)

Yup, I just had not noticed that the file already existed - sorry!

> From a TDD principle, you could 1) add a test which should fail, 2) add your code change to make the test succeeds

That is pretty much the idea with the tests I added (based on what I posted here before); the two "float64" ones fail if you run the test file with tensorflow installed from a current branch (in my case, 2.0-beta1 with GPU support installed from pip), while they succeed once the `noise.py` has been altered following the fix I submitted. Thanks for the clarification anyway!

**Review by tanzhenyu — DISMISSED** [review]

**Comment by pandrey-fr** [comment]
Some tests failed due to the `from tensorflow.python import dtypes` line. I am going to change it, but it may take a few iterations to find a way to make it work for all platforms (I do not quite get why the API is different between various python3 installations).

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by tanzhenyu — DISMISSED** [review]

**Comment by pandrey-fr** [comment]
A couple of builds failed on the pylint tests. The issue is that I respected pep 8 on 4-spaces identations, while pylint was configured to expect 2 spaces due to what appears to be the not-always-applied-and-definitely-harder-to-read-but-arguably-easier-to-write norm in tensorflow...
Should I simply update the code's indentation and have someone review the PR again?

**Comment by rthadur** [comment]
@pandrey-fr yes please fix lint errors

**Comment by pandrey-fr** [comment]
Okay, the last couple of commits should now make the PR pass the pylint tests.

On the other hand, I notices that one of the builds had actually failed on compiling a GPU op that should in no way be related to this PR - a problem that must have been fixed on the master branch for long now. So, there might be another failure when the latest commits have been reviewed, which I seemingly cannot fix but should be wiped off when merging into master, as far as I understand it.

**Comment by pandrey-fr** [comment]
On the one hand, pylint failed again on the Ubuntu Sanity build. That is because I assumed hanging indentations should be of 2 spaces, like code indentations, but actually pylint excepts 4 spaces. I am not going to argue, but this really makes very little sense to the pep-8 fanatic I tend to be. I will push a small fix later today.

On the other hand, the other builds failed on either compilation or testing issues which seem entirely unrelated to the issue... Maybe it would be worth to merge the current master into my PR branch before running tests again to do it the other way around.

**Comment by pandrey-fr** [comment]
(sorry about the pile of commits that really should have been only one - using the GitHub online code editor to go fast was not a good idea)

**Review by tanzhenyu — APPROVED** [review]
